File: Python/radixsort.py

def countingSort(arr, exp1): 

	n = len(arr) 

	output = [0] * (n) 

	count = [0] * (10) 

	for i in range(0, n): 
		index = (arr[i]/exp1) 
		count[ (index)%10 ] += 1

	for i in range(1,10): 
		count[i] += count[i-1] 

	i = n-1
	while i>=0: 
		index = (arr[i]/exp1) 
		output[ count[ (index)%10 ] - 1] = arr[i] 
		count[ (index)%10 ] -= 1
		i -= 1

	i = 0
	for i in range(0,len(arr)): 
		arr[i] = output[i] 

def radixSort(arr): 

	max1 = max(arr) 
	exp = 1
	while max1/exp > 0: 
		countingSort(arr,exp) 
		exp *= 10

radixSort(arr) 

for i in range(len(arr)): 
	print(arr[i]), 



"""
Problem : Sort the given array using radix sort.

Sort the elements by first grouping the individual digits of the same place value. Then, sort the elements according to their increasing/decreasing order.

Sample I/O:

INPUT:
121
432
564
23
1
45
788
Sorted Array in descending order is
[1, 23, 45, 121, 432, 564, 788]


Time complexity: O(nk) (worst)
Space complexity: Ω(nk) (best)
"""

File: C/Radix_Sort.c
#include <stdio.h>
//Radix Sort Uses Counting Sort as a Subroutine.
int calculateMax(int a[], int n)
{
    int max = a[0];
    for (int i = 0; i < n; i++)
    {
        if (a[i] > max)
        {
            max = a[i];
        }
    }
    return max;
}
void count_sort(int a[], int n, int pos)
{
    int count[10] = {0};
    int b[n];
    for (int i = 0; i < n; i++)
    {
        ++count[(a[i] / pos) % 10];
    }
    for (int i = 1; i <= 9; i++)
    {
        count[i] = count[i] + count[i - 1];
    }
    for (int i = n - 1; i >= 0; i--)
    {
        b[--count[(a[i] / pos) % 10]] = a[i];
    }
    for (int i = 0; i < n; i++)
    {
        a[i] = b[i];
    }
}

void radix_sort(int a[], int n)
{
    int max = calculateMax(a, n);
    for (int position = 1; max / position > 0; position *= 10)
    {
        count_sort(a, n, position);
    }
}

int main()
{
    printf("Enter the Number of Elements: ");
    int n;
    scanf("%d", &n);
    int arr[n];
    for (int i = 0; i < n; i++)
    {
        scanf("%d", &arr[i]);
    }
    radix_sort(arr, n);
    for (int i = 0; i < n; i++)
    {
        printf("%d ", arr[i]);
    }
    return 0;
}
/*
Sample Input
  10
  543 432 13 6 1 4567 45613 521 37 22
Sample Output
    1 6 13 22 37 432 521 543 4567 45613
________________________________________________
Time Complexity: O(d*(n+k)) where 'd' is the number of digits in max,k is 9
Space Complexity:O(n+k) because we are using extra space for count array
*/
